### PR TITLE
[designer] Assets picker

### PR DESCRIPTION
**Pull request by istarkov**
ref #196

- Image reverted to ordinary Image
- Added supports of "ordinary" Image `src` selection with the Asset manager. (The Idea that we save asset + prop value simultaneously, so "ordinary" controls will work without issues, i.e. in case of the image we save `prop="src" value="/imagpath" asset={...tonof info here...}`)
- Image optimizations removed (will be created in standalone package in next PR)
- Now UserProps has type https://github.com/webstudio-is/webstudio-designer/blob/c36519540c9d3493acfe0dcdd1c118ac4b6196be/packages/react-sdk/src/user-props/schema.ts#L33-L39
- Db type is: https://github.com/webstudio-is/webstudio-designer/blob/6d5cc571cb4e3bf00ec95f166a0daf9420ef50ac/packages/react-sdk/src/user-props/schema.ts#L4-L10

Next PRs:
1. Restore the Image component with optimizations at standalone package
2. Pass the Image component to SDK and use it instead of the original in case of optimizations are needed.

### CODE REVIEW

File: apps/designer/app/canvas/features/wrapper-component/wrapper-component.tsx [hunk 1]
```diff
@@ -64,7 +66,26 @@ export const WrapperComponentDev = ({
     [setSelectedElement]
   );
 
-  const userProps = useUserProps(instance.id);
+  const defaultTransformAsset = React.useCallback(
```
Comment (kof) [inline apps/designer/app/canvas/features/wrapper-component/wrapper-component.tsx:69]: lets import {useCallback} style

File: apps/designer/app/designer/features/props-panel/control.stories.tsx [hunk 2]
```diff
@@ -7,22 +7,14 @@ export default {
 
 export const Primitive: ComponentStoryObj<typeof Control> = {
   args: {
-    type: "text",
+    component: "Image",
+    userProp: { prop: "src", id: "some_id", value: "" },
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.stories.tsx:11]: we never use snake case, only in CONSTANTS_LIKE stuff, lets not even make an example lol to prevent this idea

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 3]
```diff
@@ -142,84 +220,150 @@ const RangeControl = ({
     </Flex>
   </Flex>
 );
+*/
 
 const NotImplemented = () => <div />;
 
-type PrimitiveControlProps = BaseControlProps & {
-  type: "array" | "boolean" | "date" | "number" | "object" | "text";
+const includes = <T extends string>(arr: readonly T[], v: string): v is T => {
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:227]: I think we need this as a reusable utility somewhere ... is there a good package that already has it or can we put it somewhere else?

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 4]
```diff
@@ -142,84 +220,150 @@ const RangeControl = ({
...
+  return arr.includes(v as never);
 };
 
-type ColorControlProps = BaseControlProps & {
-  type: "color";
-  presetColors?: Array<string>;
+const isStringArray = (arr: unknown): arr is Array<string> => {
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:231]: this could be also a good utility in the same belt

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 5]
```diff
@@ -142,84 +220,150 @@ const RangeControl = ({
...
-    | "check"
-    | "inline-check"
-    | "select"
-    | "multi-select";
-  options: Array<string>;
-};
+  const defaultValue = argType?.control.defaultValue;
+  const type = argType?.control.type;
+
+  if (type == null) {
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:255]: can type be null and undefined? if yes then its correct

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 6]
```diff
@@ -142,84 +220,150 @@ const RangeControl = ({
...
+    warnOnce(
+      type == null,
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:257]: that's just true
```suggestion
      true,
```

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 7]
```diff
@@ -142,84 +220,150 @@ const RangeControl = ({
...
 
-export type ControlProps =
-  | PrimitiveControlProps
-  | OptionsControlProps
-  | FileControlProps
-  | ColorControlProps
-  | RangeControlProps;
+  if (typeof type !== "string") {
+    warnOnce(
+      typeof type !== "string",
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:265]:
```suggestion
      true,
```

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 8]
```diff
@@ -142,84 +220,150 @@ const RangeControl = ({
...
+        value={value}
+        defaultValue={defaultValue}
+        onChange={onChangePropValue}
+      />
+    );
+  }
+
+  if (
+    includes(
+      [...radioControlTypes, ...checkboxControlTypes, ...selectControlTypes],
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:309]: this merge can be done once in the module scope

File: apps/designer/app/designer/features/props-panel/props-panel.tsx [hunk 9]
```diff
@@ -200,8 +210,14 @@ export const PropsPanel = ({
       <PlusIcon />
     </Button>
   );
+
   return (
-    <>
+    <Box
+      css={{
+        position:
+          "relative" /* Hack - value picker popover positioning depends on it */,
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/props-panel.tsx:218]: now we gotta fix it once for both cases, I think we should have that container in a location for both, seems like the closest shared folder is designer/shared/inspector/pane.tsx

File: apps/designer/app/designer/features/props-panel/use-props-logic.ts [hunk 10]
```diff
@@ -19,11 +19,16 @@ declare module "~/shared/pubsub" {
...
-  id: UserProp["id"],
-  field: "prop" | "value",
-  value: UserProp["prop"] | UserProp["value"]
-) => void;
+// First infer is used to iterate over Union type
+type UserPropValue = UserProp extends infer IterUserProp
+  ? IterUserProp extends { value: infer Value }
+    ? Value
+    : never
+  : never;
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/use-props-logic.ts:27]: almost too complex type for my test, please check if it can be simplified some how

File: apps/designer/app/designer/features/props-panel/use-props-logic.ts [hunk 11]
```diff
@@ -19,11 +19,16 @@ declare module "~/shared/pubsub" {
...
+
+type handleChangePropName = (id: UserProp["id"], name: string) => void;
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/use-props-logic.ts:29]: types are allways capitalized

File: packages/asset-uploader/src/env.ts [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export const env = {
+  VERCEL_ENV: process.env.VERCEL_ENV,
```
Comment (kof) [inline packages/asset-uploader/src/env.ts:2]: can we please abstract away VERCEL_ENV and use something that will not change no matter if we use vercel or something else?

File: packages/asset-uploader/src/utils/format-asset.test.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+  projectId: "id",
+  size: 2135,
+  format: "png",
+  createdAt: new Date(),
+  description: "",
+  meta: JSON.stringify({ width: 200, height: 200 }),
+};
+
+describe("getAssetPath", () => {
+  const OLD_ENV = process.env;
```
Comment (kof) [inline packages/asset-uploader/src/utils/format-asset.test.ts:18]: `OLD_ENV` where is this coming from?

Comment (kof) [inline packages/asset-uploader/src/utils/format-asset.test.ts:18]: I got confused for a second, we use SCREAMING_CASE only when we need to export constants like a config, not when its a local constant

File: packages/asset-uploader/src/utils/format-asset.test.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    };
+  });
+
+  afterAll(() => {
+    process.env = OLD_ENV;
+  });
+
+  describe("local", () => {
+    test("returns none on Location.FS if VERCEL_ENV not set", () => {
+      process.env.VERCEL_ENV = undefined;
```
Comment (kof) [inline packages/asset-uploader/src/utils/format-asset.test.ts:33]: do you like the mocking more than passing env as a second argument that defaults to process.env? It's clean and doesn't depend on jest magic

File: packages/asset-uploader/src/utils/format-asset.test.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+      const { formatAsset } = require("./format-asset");
```
Comment (kof) [inline packages/asset-uploader/src/utils/format-asset.test.ts:35]: the above suggestion would also prevent the need for require

File: packages/asset-uploader/src/utils/format-asset.ts [hunk 16]
```diff
@@ -24,8 +25,19 @@ export const formatAsset = (asset: DbAsset): Asset => {
     };
   }
 
+  let transform: ImageAsset["transform"] = "none";
+
+  if (asset.location === "REMOTE") {
+    transform = "relative";
+
+    if (env.VERCEL_ENV != null) {
```
Comment (kof) [inline packages/asset-uploader/src/utils/format-asset.ts:33]: same here, how can we have a flag that is independent of the specific environment name like "vercel", "cloudflare" or whatever

File: packages/project/src/db/props.ts [hunk 17]
```diff
@@ -1,22 +1,75 @@
...
 import { prisma } from "@webstudio-is/prisma-client";
+import { formatAsset } from "@webstudio-is/asset-uploader/server";
 
 export const loadByTreeId = async (treeId: Tree["id"]) => {
   const instancePropsEntries = await prisma.instanceProps.findMany({
     where: { treeId },
   });
 
+  // Fin all asset_id in all props
+  const assets_id: string[] = [];
```
Comment (kof) [inline packages/project/src/db/props.ts:19]: oh nooo, now snake landed in code and I had to create an issue for another eslint rule https://github.com/webstudio-is/webstudio-designer/issues/445 which is now yours 🙃

File: packages/project/src/db/props.ts [hunk 18]
```diff
@@ -1,22 +1,75 @@
...
+
+  for (const instanceProps of instancePropsEntries) {
+    const props = JSON.parse(instanceProps.props);
+    const dpProps = zUserDbProps.parse(props);
+
+    for (const dbProp of dpProps) {
+      if ("asset_id" in dbProp) {
```
Comment (kof) [inline packages/project/src/db/props.ts:26]: same

File: packages/project/src/db/props.ts [hunk 19]
```diff
@@ -1,22 +1,75 @@
...
   const instancePropsEntries = await prisma.instanceProps.findMany({
     where: { treeId },
   });
 
+  // Fin all asset_id in all props
+  const assets_id: string[] = [];
+
+  for (const instanceProps of instancePropsEntries) {
+    const props = JSON.parse(instanceProps.props);
+    const dpProps = zUserDbProps.parse(props);
```
Comment (kof) [inline packages/project/src/db/props.ts:23]: zod schema always starts capital letter, just like its type

File: packages/project/src/db/props.ts [hunk 20]
```diff
@@ -1,22 +1,75 @@
...
+          resProps.push({
+            id: dbProp.id,
+            prop: dbProp.prop,
+            asset,
+            required: dbProp.required,
+          });
+        } else {
+          // Can be ok if asset was deleted having that no db constraint checks exists
+          // TODO: remove throw on something else
+          throw new Error(`asset with ${dbProp.asset_id} not found`);
```
Comment (kof) [inline packages/project/src/db/props.ts:63]: just fui for consistent error style
```suggestion
          throw new Error(`Asset with assetId "${dbProp.asset_id}" not found`);
```

File: packages/react-sdk/src/component-utils/image-loaders.ts [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+ * https://developers.cloudflare.com/images/image-resizing/url-format/
+ **/
+export const cloudflareRelativeImageLoader: ImageLoader = ({
+  width,
+  src,
+  quality,
+}) => {
+  if (process.env.NODE_ENV !== "production") {
+    warnOnce(
+      !allSizes.includes(width),
```
Comment (kof) [inline packages/react-sdk/src/component-utils/image-loaders.ts:15]: easier to scan, please trust me on this

```suggestion
      allSizes.includes(width) === false,
```

File: packages/react-sdk/src/component-utils/image-loaders.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
+import warnOnce from "warn-once";
+import { allSizes, type ImageLoader } from "./image";
+
+/**
+ * Default image loader in case of no loader provided
+ * https://developers.cloudflare.com/images/image-resizing/url-format/
+ **/
+export const cloudflareRelativeImageLoader: ImageLoader = ({
```
Comment (kof) [inline packages/react-sdk/src/component-utils/image-loaders.ts:8]: just realized that normal open source package user will not have cloudflare installed, so they don't need this cloudflare logic to be there at all, what would be a better place for CF specific stuff?

File: packages/react-sdk/src/component-utils/image-loaders.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+/**
+ * Same as cloudflareRelativeImageLoader just uses full URL for tranform path
+ */
+export const cloudflareRemoteImageLoader: ImageLoader = (props) => {
+  // Cloudflare docs say that we don't need to urlencode the path params
+  const pathname = cloudflareRelativeImageLoader(props);
+
+  const resizeOrigin = "https://webstudio.is/";
```
Comment (kof) [inline packages/react-sdk/src/component-utils/image-loaders.ts:33]: can't be hardcoded

File: packages/react-sdk/src/components/component-type.ts [hunk 24]
```diff
@@ -2,6 +2,7 @@ import type { Style } from "../css";
 import { z } from "zod";
 import React from "react";
 import { IconProps } from "@webstudio-is/icons";
+import { TransformAsset } from "~/user-props";
```
Comment (kof) [inline packages/react-sdk/src/components/component-type.ts:5]: use type
```suggestion
import type { TransformAsset } from "~/user-props";
```

File: packages/react-sdk/src/components/component-type.ts [hunk 25]
```diff
@@ -31,6 +33,7 @@ export const WsComponentMeta = z.lazy(() =>
       isListed: z.boolean(),
       label: z.string(),
       children: z.optional(z.array(z.string())),
+      transformAsset: z.optional(z.any()),
```
Comment (kof) [inline packages/react-sdk/src/components/component-type.ts:36]: any???

File: packages/react-sdk/src/components/image.ws.tsx [hunk 26]
```diff
@@ -17,6 +23,26 @@ const defaultStyle = {
...
+
+const ensureBoolean = (value: unknown): boolean | undefined => {
+  if (typeof value === "boolean") return value;
+  return undefined;
+};
+
+const assertUnreachable = (_arg: never) => {
+  throw new Error("Didn't expect to get here");
+};
+
```
Comment (kof) [inline packages/react-sdk/src/components/image.ws.tsx:45]: looks like a set of useful generic utils like in some other case before

File: packages/react-sdk/src/components/image.ws.tsx [hunk 27]
```diff
@@ -26,6 +52,44 @@ const meta: WsComponentMeta<typeof Image> = {
   isInlineOnly: false,
   isListed: true,
   label: "Image",
+  transformAsset: (assetProp, props) => {
```
Comment (kof) [inline packages/react-sdk/src/components/image.ws.tsx:55]: now that makes meta non-serializable, not sure this is not a problem

File: packages/react-sdk/src/components/image.ws.tsx [hunk 28]
```diff
@@ -26,6 +52,44 @@ const meta: WsComponentMeta<typeof Image> = {
...
+    }
+
+    const { asset } = assetProp;
+
+    let loader = localImageLoader;
+
+    if ("transform" in asset) {
+      switch (asset.transform) {
+        case "relative":
+          loader = cloudflareRelativeImageLoader;
```
Comment (kof) [inline packages/react-sdk/src/components/image.ws.tsx:67]: again, image has no idea of cloudflare

Comment (kof) [reply packages/react-sdk/src/components/image.ws.tsx:67]: I think ultimately this entire transform belongs somewhere else

**Review by kof — CHANGES_REQUESTED** [review]
good initial draft, lets clean it up

File: packages/react-sdk/src/components/image.ws.tsx (hunk 27 above)
Comment (kof) [reply packages/react-sdk/src/components/image.ws.tsx:55]: discussed: we want a universal transformers interface that can be sticked together like middleware. This one is not universal right now

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 4 above)
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx:231]: Disagree, people even don't reuse npm packages already exists in big repos, and reusing a one-liner like this is nearly impossible. I think I wrote string array type guard for the first time in my life, and the same for includes. They are rare, the main reason is to be type guards, and not every piece of code should has own module or be in some utils. Even so - reusability is near impossible.

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 5 above)
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx:255]: it can, its even `any` as control is not defined on argtype

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 4 above)
Comment (kof) [reply apps/designer/app/designer/features/props-panel/control.tsx:231]: yeah maybe this one is a bit too special

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 29]
```diff
@@ -117,109 +160,186 @@ const BooleanControl = ({
...
+      `Control type for prop: ${
+        userProp.prop
+      } component: ${component} is not a string but ${typeof type}`
+    );
+
+    return <NotImplemented />;
+  }
+
+  if (isFeatureEnabled("assets")) {
+    if (component === "Image" && userProp.prop === "src") {
```
Comment (rpominov) [inline apps/designer/app/designer/features/props-panel/control.tsx:254]: Why we check `component` and `userProp.prop` directly rather than going via `const type = argType?.control.type` like for all other control types? Is this temporary?

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 30]
```diff
@@ -117,109 +160,186 @@ const BooleanControl = ({
...
+  return Array.isArray(arr) && arr.every((v) => typeof v === "string");
 };
 
-type RangeControlProps = BaseControlProps<number> & {
-  type: "range";
-  min: number;
-  max: number;
-  step: number;
+const assertUnreachable = (_arg: never) => {
+  throw new Error("Didn't expect to get here");
```
Comment (rpominov) [inline apps/designer/app/designer/features/props-panel/control.tsx:205]: Would be good to include `_arg` in the message. Just in case :)

File: packages/asset-uploader/src/utils/get-asset-path.ts [hunk 31]
```diff
@@ -8,18 +8,28 @@ const fsEnv = FsEnv.parse(process.env);
...
+    );
+
+    return locUrl.pathname;
   }
 
   if (asset.location === Location.REMOTE && s3Envs.success) {
     if (s3Envs.data.ASSET_CDN_URL) {
       const cndUrl = new URL(s3Envs.data.ASSET_CDN_URL);
-      cndUrl.pathname = encodeURIComponent(asset.name);
+      cndUrl.pathname = encodeURIComponent(encodeURIComponent(asset.name));
```
Comment (rpominov) [inline packages/asset-uploader/src/utils/get-asset-path.ts:27]: encodeURIComponent twice? Same below.

File: packages/react-sdk/src/prop-transforms/image.ts [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+        quality: ensureStringOrNumber(imageValueProps.quality),
+        loader,
+        // User can overwrite optimize
+        optimize: ensureBoolean(imageValueProps.optimize) ?? true,
+      });
+
+      if (imageAttributes != null) {
+        for (const [key, value] of Object.entries(imageAttributes)) {
+          res.push({
+            id: `imgt_${gId++}`,
```
Comment (rpominov) [inline packages/react-sdk/src/prop-transforms/image.ts:77]: What `imgt` means and what `g` means in `gId`? 😅

File: apps/designer/app/canvas/canvas.tsx [hunk 33]
```diff
@@ -129,6 +130,13 @@ export const Canvas = ({ data }: CanvasProps): JSX.Element | null => {
   const isPreviewMode = useSubscribePreviewMode();
   const elements = useElementsTree();
 
+  propTransforms.initComponentPropsTransformer(
+    "Image",
+    propTransforms.Image.createPropsTransformer(
+      data.params?.propTransforms?.Image ?? null
+    )
+  );
```
Comment (kof) [inline apps/designer/app/canvas/canvas.tsx:138]: should be more generic, for any component

File: apps/designer/app/routes/$.tsx [hunk 34]
```diff
@@ -76,7 +76,16 @@ export const loader: LoaderFunction = async ({
       throw json("Page not found", { status: 404 });
     }
 
-    return { ...canvasData, env, mode };
+    const params = { ...canvasData.params };
```
Comment (kof) [inline apps/designer/app/routes/$.tsx:79]: why do we need `params` and not canvasData.propTransform directly?

File: apps/designer/app/routes/$.tsx [hunk 35]
```diff
@@ -76,7 +76,16 @@ export const loader: LoaderFunction = async ({
...
+
+    if (env.RESIZE_ORIGIN != null) {
+      params.propTransforms = params.propTransforms ?? {};
+      params.propTransforms.Image = {
+        resizeOrigin: env.RESIZE_ORIGIN,
+      };
+    }
```
Comment (kof) [inline apps/designer/app/routes/$.tsx:86]: these are technically not props, these are transformer configs, we configure a middleware, then the middleware will receive component props and know what to do with it.

Comment (kof) [reply apps/designer/app/routes/$.tsx:86]: I think this datastructure makes more sense to me, wdyt?

```
canvasData.propTransforms = [
  {
    component: 'Image',
    config:  [resizeOrigin: env.RESIZE_ORIGIN,}
  }
]
```

Comment (kof) [reply apps/designer/app/routes/$.tsx:86]: but looking at his all makes me wonder if we could do what you suggested initial - flat config for all transformers, transformer takes what it needs. In the worst case we can make this here work later, right?

File: packages/asset-uploader/src/utils/get-asset-path.ts (hunk 31 above)
Comment (istarkov) [reply packages/asset-uploader/src/utils/get-asset-path.ts:27]: Yep otherwise spaces, cyrillic etc is not working. 
On upload we do following https://github.com/webstudio-is/webstudio-designer/blob/6488305087e661923415971a36d491727e336072/packages/asset-uploader/src/targets/s3/upload.ts#L85 (otherwise s3 will respond with error I think)

Then we need to encode filename into a path, but filename is already `filename = encodeURIComponent(asset.name)` see link above
Having that here we need `encodeUriComponent(filename)` we got 2nd encode ;-)

File: packages/react-sdk/src/prop-transforms/image.ts (hunk 32 above)
Comment (istarkov) [reply packages/react-sdk/src/prop-transforms/image.ts:77]: Just need some id non intersected with others. 
g means global ;-) If we have some convention about mod level state vars tell me. 
Usualy I use underscore after var, i.e. `id_`

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 30 above)
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx:205]: Its typescript hack, see `_arg: never` so any call to this function with any non `never` argument would cause ts error. `throw new Error` just instead of comment. IMO no need to check typescript in runtime adding any logic on never arg or output it

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 29 above)
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx:254]: Because we want this only for selected component `Image`, for known prop name 'src' only.

`type` there is coming from storybook and is `string` or like, there are no type `asset` or like on our components.

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 36]
```diff
@@ -117,109 +160,186 @@ const BooleanControl = ({
...
+    );
+    return <NotImplemented />;
+  }
+
+  if (typeof type !== "string") {
+    warnOnce(
+      true,
+      `Control type for prop: ${
+        userProp.prop
+      } component: ${component} is not a string but ${typeof type}`
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx:247]: `Control type for prop:` blabla `is not a string` will fix

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 30 above)
Comment (rpominov) [reply apps/designer/app/designer/features/props-panel/control.tsx:205]: I mean what if a user reports a bug saying they're getting a error "Didn't expect to get here"? It would be better for us if user comes with something like "Didn't expect to get here. Got: something".

File: packages/asset-uploader/src/utils/get-asset-path.ts (hunk 31 above)
Comment (rpominov) [reply packages/asset-uploader/src/utils/get-asset-path.ts:27]: Ok, just wanted to make sure this is not a mistake

File: packages/react-sdk/src/prop-transforms/image.ts (hunk 32 above)
Comment (rpominov) [reply packages/react-sdk/src/prop-transforms/image.ts:77]: Can we do something like `image_transform_${globalId}` so it would be less confusing? (if I guessed correctly what `imgt` means)

File: packages/asset-uploader/src/utils/get-asset-path.ts (hunk 31 above)
Comment (istarkov) [reply packages/asset-uploader/src/utils/get-asset-path.ts:27]: Ill add my comments into a code as its really confusing and probably need some investigations, like storing fileName on asset

File: packages/react-sdk/src/prop-transforms/image.ts (hunk 32 above)
Comment (istarkov) [reply packages/react-sdk/src/prop-transforms/image.ts]: Done

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 30 above)
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx]: Moved msg out of scope. Yep found a case that it can be executed in runtime - for example, if a developer will forget to set `return` in the code below, it would be no ts error.

File: apps/designer/app/designer/features/props-panel/control.tsx (hunk 36 above)
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/designer/app/designer/features/props-panel/control.tsx]: fixed

File: apps/designer/app/routes/$.tsx (hunk 35 above)
Comment (istarkov) [reply apps/designer/app/routes/$.tsx:83]: Ok transforms now are in a standalone folder, and just transforms https://github.com/webstudio-is/webstudio-designer/blob/8539fbf2f41d4763ff308da662106ed843164c88/packages/react-sdk/src/prop-transforms/index.ts#L1

In the future, it can be an argument to sdk components.

All logic to create transforms etc are  moved here 

https://github.com/webstudio-is/webstudio-designer/blob/32b33b2d176f5687d8c2a28b42663dc5b5cdcee9/packages/react-sdk/src/user-props/prop-transforms.ts#L27

And now more generic.

https://github.com/webstudio-is/webstudio-designer/blob/32b33b2d176f5687d8c2a28b42663dc5b5cdcee9/packages/react-sdk/src/tree/root.ts#L36

File: packages/react-sdk/src/user-props/prop-transforms.ts [hunk 37]
```diff
@@ -0,0 +1,64 @@
+import warnOnce from "warn-once";
+import type * as components from "../components";
+import type { TransformProps } from "~/user-props";
+import type * as propTransforms from "~/prop-transforms";
+
+// Bad rule, in modern ts {} is good type see https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#improved-intersection-reduction-union-compatibility-and-narrowing
+// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (rpominov) [inline packages/react-sdk/src/user-props/prop-transforms.ts:7]: I had a case where this rule was in the way too, and `{}` worked perfectly

**Review by rpominov — APPROVED** [review]

File: packages/react-sdk/src/components/meta.ts [hunk 38]
```diff
@@ -32,14 +32,32 @@ const meta = {
...
-type MetaProps = Record<
-  string,
-  {
-    type: string;
-    required: boolean;
-    defaultValue: null | string;
-    options?: string[];
-  }
->;
+type MetaProp =
```
Comment (istarkov) [inline packages/react-sdk/src/components/meta.ts:35]: In case of we switch to `ts` instead of `json` generation.
I.e. `as const`, I would be able to generate this types from ts.

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 39]
```diff
@@ -117,107 +145,170 @@ const BooleanControl = ({
...
-  presetColors?: Array<string>;
+const ImageControl = ({ asset, onChange }: ImageControlProps) => {
+  return (
+    <ValuePickerPopover
+      title="Images"
+      content={
+        <ImageManager
+          onChange={(asset) => {
+            onChange(asset);
+          }}
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:161]:
```suggestion
          onChange={onChange}
```

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 40]
```diff
@@ -117,107 +145,170 @@ const BooleanControl = ({
...
+          onChange={onChangePropValue}
+          options={options ?? DEFAULT_OPTIONS}
+          type={type}
+        />
+      );
+    }
+
+    assertUnreachable(
+      type,
+      `Did't expect to be here, unknown control type ${type}`
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:304]: funny, but
```suggestion
      `Unknown control type ${type}`
```

File: apps/designer/app/designer/features/props-panel/control.tsx [hunk 41]
```diff
@@ -117,107 +145,170 @@ const BooleanControl = ({
...
+    );
+  }
+
+  warnOnce(
+    true,
+    `Control type ${type} not implemented for prop: ${userProp.prop} component: ${component}`
```
Comment (kof) [inline apps/designer/app/designer/features/props-panel/control.tsx:310]:
```suggestion
    `Control type "${type}" is not implemented for prop: "${userProp.prop}" in component "${component}"`
```

File: packages/project/src/db/props.ts [hunk 42]
```diff
@@ -1,22 +1,73 @@
 import {
   type Tree,
   type AllUserProps,
-  UserProps,
+  type UserProp,
+  // TODO: move zod schema into project package
```
Comment (kof) [inline packages/project/src/db/props.ts:5]: forgotten? or wanted to do with a separate PR? 
inn any case we use the `@todo` style comment

File: packages/project/src/db/props.ts [hunk 43]
```diff
@@ -1,22 +1,73 @@
...
   const instancePropsEntries = await prisma.instanceProps.findMany({
     where: { treeId },
   });
 
+  // Fin all assetId in all props
+  const assetIds: string[] = [];
+
+  for (const instanceProps of instancePropsEntries) {
+    const props = JSON.parse(instanceProps.props);
+    const dpProps = UserDbProps.parse(props);
```
Comment (kof) [inline packages/project/src/db/props.ts:23]: just call it parsedProps
```suggestion
    const parsedProps = UserDbProps.parse(props);
```

File: packages/project/src/db/props.ts [hunk 44]
```diff
@@ -1,22 +1,73 @@
...
+
   return instancePropsEntries.map((instanceProps) => {
     const props = JSON.parse(instanceProps.props);
-    UserProps.parse(props);
+    const resProps: UserProp[] = [];
+    // Can be changed onto `props as ` having that above is already checked, no optmizations now
+    const dpProps = UserDbProps.parse(props);
+
+    for (const dbProp of dpProps) {
+      if (dbProp.assetId != null) {
```
Comment (kof) [inline packages/project/src/db/props.ts:51]: can we please keep the nesting level low?
```suggestion
      if (dbProp.assetId == null) {
         resultProps.push(parsedProp)
         continue
      }
```

File: packages/project/src/db/props.ts [hunk 45]
```diff
@@ -1,22 +1,73 @@
...
+  const assetsMap = new Map(
+    assets.map((asset) => [asset.id, formatAsset(asset)])
+  );
+
   return instancePropsEntries.map((instanceProps) => {
     const props = JSON.parse(instanceProps.props);
-    UserProps.parse(props);
+    const resProps: UserProp[] = [];
+    // Can be changed onto `props as ` having that above is already checked, no optmizations now
+    const dpProps = UserDbProps.parse(props);
```
Comment (kof) [inline packages/project/src/db/props.ts:48]:
```suggestion
    const parsedProps = UserDbProps.parse(props);
```

File: packages/project/src/db/props.ts [hunk 46]
```diff
@@ -1,22 +1,73 @@
...
+
+    for (const dbProp of dpProps) {
+      if (dbProp.assetId != null) {
+        const asset = assetsMap.get(dbProp.assetId);
+        if (asset) {
+          resProps.push({
+            ...dbProp,
+            asset,
+          });
+        } else {
```
Comment (kof) [inline packages/project/src/db/props.ts:58]:
```suggestion
          continue;
        } 
```

File: packages/project/src/db/props.ts (hunk 43 above)
Comment (kof) [reply packages/project/src/db/props.ts:23]: cleaned it up for you

```ts
return instancePropsEntries.map((instanceProps) => {
  const props = JSON.parse(instanceProps.props);
  // Can be changed onto `props as ` having that above is already checked, no optmizations now
  const parsedProps = UserDbProps.parse(props);
  const resultProps: UserProp[] = [];

  for (const parsedProp of parsedProps) {
    if (parsedProp.assetId == null) {
      resultProps.push(parsedProp);
      continue;
    }
    
    const asset = assetsMap.get(dbProp.assetId);
    if (asset) {
      resProps.push({
        ...dbProp,
        asset,
      });
      continue
    } 
    // Can be ok if asset was deleted having that no db constraint checks exists
    // @todo: remove throw on something else
    throw new Error(`Asset with assetId "${dbProp.assetId}" was not found`);
  }

  return {
    ...instanceProps,
    props: resultProps,
  };
});
```

File: packages/project/src/db/props.ts [hunk 47]
```diff
@@ -1,22 +1,73 @@
...
+          // Can be ok if asset was deleted having that no db constraint checks exists
+          // TODO: remove throw on something else
+          throw new Error(`Asset with assetId "${dbProp.assetId}" not found`);
```
Comment (kof) [inline packages/project/src/db/props.ts:61]: Deleted image that is still being referenced from the props is a normal thing, we need to handle it gratefully without breaking

**Review by kof — CHANGES_REQUESTED** [review]
almost there, hang on

File: packages/project/src/db/props.ts (hunk 42 above)
Comment (istarkov) [reply packages/project/src/db/props.ts:5]: Separate

File: packages/project/src/db/props.ts (hunk 43 above)
Comment (istarkov) [reply packages/project/src/db/props.ts:23]: IMO dbProps is perfectly fine as gives information that this props are somehow different from userProps.
Changed resultProps onto userProps to make it more visible

File: packages/project/src/db/props.ts (hunk 45 above)
Comment (istarkov) [reply packages/project/src/db/props.ts:48]: db is better name, we have different UserProps representation on UI and DB side

**Review by kof — APPROVED** [review]
